packages/notifications/push/amplify_push_notifications/android/src/main/kotlin/com/amazonaws/amplify/amplify_push_notifications/PushNotificationEventsStreamHandler.kt

```diff
@@ -0,0 +1,178 @@
+
+    val eventName: String
+        get() = when (this) {
+            TOKEN_RECEIVED -> "TOKEN_RECEIVED"
+            NOTIFICATION_OPENED -> "NOTIFICATION_OPENED"
+            LAUNCH_NOTIFICATION_OPENED -> "LAUNCH_NOTIFICATION_OPENED"
+            FOREGROUND_MESSAGE_RECEIVED -> "FOREGROUND_MESSAGE_RECEIVED"
+            BACKGROUND_MESSAGE_RECEIVED -> "BACKGROUND_MESSAGE_RECEIVED"
+            ERROR -> "ERROR"
+        }
+
+    val eventChannelName: String
+        get() = "${PushNotificationPluginConstants.CHANNEL_NAME_PREFIX}/event/$eventName"
+}
+
+@InternalAmplifyApi
+data class PushNotificationsEvent(
+    val event: NativeEvent, val payload: Map<Any, Any?>
+)
+
+@InternalAmplifyApi
+class PushNotificationEventsStreamHandler constructor(
+    private val associatedNativeEvent: NativeEvent
+) : EventChannel.StreamHandler {
+
+    private var eventChannel: EventChannel? = null
+    private var eventSink: EventSink? = null
+
+    fun initEventChannel(binaryMessenger: BinaryMessenger) {
+        eventChannel = EventChannel(
+            binaryMessenger, associatedNativeEvent.eventChannelName
+        )
+        eventChannel?.setStreamHandler(this)
+    }
+
+    fun deInitEventChannel() {
+        eventChannel = null
+    }
+
+    override fun onListen(arguments: Any?, sink: EventSink?) {
+        eventSink = sink
+        flushEvents()
+    }
+
+    override fun onCancel(arguments: Any?) {
+        eventSink = null
+        eventQueue.clear()
+    }
+
+    private val eventQueue = mutableListOf<PushNotificationsEvent>()
+
+    fun send(payload: Map<Any, Any?>) {
+        val event = PushNotificationsEvent(associatedNativeEvent, payload)
+        eventSink?.success(payload) ?: run {
+            eventQueue.add(event)
+        }
+    }
+
+    fun sendError(exception: Exception) {
+        val exceptionMap = mapOf<Any, Any?>(
+            "associatedNativeEventName" to associatedNativeEvent.eventName,
+            "message" to exception.message,
+            "details" to null
+        )
+        eventSink?.error(
+            exceptionMap["associatedNativeEventName"] as String,
+            exceptionMap["message"] as String,
+            exceptionMap["details"]
+        ) ?: run {
+            eventQueue.add(PushNotificationsEvent(NativeEvent.ERROR, exceptionMap))
+        }
+    }
+
+    private fun flushEvents() {
+        try {
+            eventSink?.let {
+                while (eventQueue.isNotEmpty()) {
+                    val eventFromQueue = eventQueue.removeFirst()
+                    // Check if it is an Error event and handle accordingly by using .error method
+                    if (eventFromQueue.event.eventName == NativeEvent.ERROR.eventName) {
+                        val exception = eventFromQueue.payload
+                        it.error(
+                            exception["associatedNativeEventName"] as String?,
```

Comment:
This should be a data class ideally - using a map with Strings is super error-prone